numba/cuda/tests/cudapy/test_overload.py

```diff
@@ -0,0 +1,290 @@
+    def check_overload(self, kernel, expected):
+        x = np.ones(1, dtype=np.int32)
+        cuda.jit(kernel)[1, 1](x)
+        self.assertEqual(x[0], expected)
+
+    def test_generic(self):
+        def kernel(x):
+            generic_func_1(x)
+
+        expected = GENERIC_FUNCTION_1
+        self.check_overload(kernel, expected)
+
+    def test_cuda(self):
+        def kernel(x):
+            cuda_func_1(x)
+
+        expected = CUDA_FUNCTION_1
+        self.check_overload(kernel, expected)
+
+    def test_generic_and_cuda(self):
+        def kernel(x):
+            generic_func_1(x)
+            cuda_func_1(x)
+
+        expected = GENERIC_FUNCTION_1 * CUDA_FUNCTION_1
+        self.check_overload(kernel, expected)
+
+    def test_call_two_generic_calls(self):
+        def kernel(x):
+            generic_func_1(x)
+            generic_func_2(x)
+
+        expected = GENERIC_FUNCTION_1 * GENERIC_FUNCTION_2
+        self.check_overload(kernel, expected)
+
+    def test_call_two_cuda_calls(self):
+        def kernel(x):
+            cuda_func_1(x)
+            cuda_func_2(x)
+
+        expected = CUDA_FUNCTION_1 * CUDA_FUNCTION_2
+        self.check_overload(kernel, expected)
+
+    def test_generic_calls_generic(self):
+        def kernel(x):
+            generic_calls_generic(x)
+
+        expected = GENERIC_CALLS_GENERIC * GENERIC_FUNCTION_1
+        self.check_overload(kernel, expected)
+
+    def test_generic_calls_cuda(self):
+        def kernel(x):
+            generic_calls_cuda(x)
+
+        expected = GENERIC_CALLS_CUDA * CUDA_FUNCTION_1
+        self.check_overload(kernel, expected)
+
+    def test_cuda_calls_generic(self):
+        def kernel(x):
+            cuda_calls_generic(x)
+
+        expected = CUDA_CALLS_GENERIC * GENERIC_FUNCTION_1
+        self.check_overload(kernel, expected)
+
+    def test_cuda_calls_cuda(self):
+        def kernel(x):
+            cuda_calls_cuda(x)
+
+        expected = CUDA_CALLS_CUDA * CUDA_FUNCTION_1
+        self.check_overload(kernel, expected)
+
+    def test_call_hardware_overloaded(self):
+        def kernel(x):
+            hardware_overloaded(x)
+
+        expected = CUDA_HARDWARE_OL
+        self.check_overload(kernel, expected)
+
+    def test_generic_calls_hardware_overloaded(self):
+        def kernel(x):
+            generic_calls_hardware_overloaded(x)
+
+        expected = GENERIC_CALLS_HARDWARE_OL * CUDA_HARDWARE_OL
+        self.check_overload(kernel, expected)
+
+    def test_cuda_calls_hardware_overloaded(self):
+        def kernel(x):
+            cuda_calls_hardware_overloaded(x)
+
+        expected = CUDA_CALLS_HARDWARE_OL * CUDA_HARDWARE_OL
+        self.check_overload(kernel, expected)
+
+    def test_hardware_overloaded_calls_hardware_overloaded(self):
+        def kernel(x):
+            hardware_overloaded_calls_hardware_overloaded(x)
+
+        expected = CUDA_HARDWARE_OL_CALLS_HARDWARE_OL * CUDA_HARDWARE_OL
+        self.check_overload(kernel, expected)
```

Comment:
In this section, maybe test that the CPU calls would pick up the "generic" variants?

Reply:
Added.